draft-ietf-mls-protocol.md

```diff
@@ -1524,223 +1460,294 @@ The creator of the group constructs an Init message as follows:
+accurate, because each node can only compute the secret and private
+key for nodes in its direct path.  This creates the possibility
+that a malicious participant could cause a denial of service by sending a
+handshake message with invalid values for public keys in the ratchet
+tree. ]]
 
-* Blank the path from the removed leaf to the root node for
-  the time of the computation
-* Truncate the tree such that the rightmost non-blank leaf is the
-  last node of the tree, for the time of the computation
-* Generate a fresh leaf key pair
-* Compute its direct path in the current ratchet tree, starting from
-  the sender's leaf
+### Welcoming New Members
 
-A member receiving a Remove message first verifies
-the signature on the message.  The member then updates its
-state as follows:
+The sender of a Commit message is responsible for sending Welcome messages to
+any new members added via Add proposals.  The Welcome message provides the new
+member with the current state of the group, after the application of the Commit
+message.  So the new member will not be able to decrypt or verify the Commit
+message, but will have the secrets it needs to participate in the epoch
+initiated by the Commit message.
 
-* Update the ratchet tree by setting to blank all nodes in the
-  direct path of the removed leaf, and also setting the root node
-  to blank
-* Truncate the tree such that the rightmost non-blank leaf is the
-  last node of the tree
-* Update the ratchet tree by replacing nodes in the direct
-  path from the sender's leaf using the information in the Remove message
+The information in a Welcome message is encrypted for the new member using HPKE.
+The recipient key pair for the HPKE encryption is the one included in the
+indicated ClientInitKey.
 
-Note that there must be at least one non-null element in
-the tree, since any valid GroupContext must have the current member in
-the tree and self-removal is prohibited
+~~~~~
+struct {
+    HPKEPublicKey public_key;
+    optional<Credential> credential;
+} RatchetNode;
 
-The `update_secret` resulting from this change is the `path_secret[i+1]`
-derived from the `path_secret[i]` associated to the root node.
+struct {
+    ProtocolVersion version = mls10;
+    opaque group_id<0..255>;
+    uint32 epoch;
+    optional<RatchetNode> tree<1..2^32-1>;
+    uint32 index;
+    opaque interim_transcript_hash<0..255>;
+    opaque epoch_secret<0..255>;
+} WelcomeInfo;
+
+struct {
+    opaque client_init_key_id<0..255>;
+    HPKECiphertext encrypted_welcome_info;
+} Welcome;
+~~~~~
+
+In the description of the tree as a list of nodes, the `credential`
+field for a node MUST be populated if and only if that node is a
+leaf in the tree.
 
+[[ OPEN ISSUE: If multiple members are being added, it would be nice to be able
+to encrypt the group state once and just encrypt the key to the members.  In
+that case, you would also need to factor out the "index" field. ]]
```

Comment:
There's a section below here, "Merging Updates", that I think you can remove. Merging multiple Commits doesn't make sense